File: AN_2/SEMESTRUL_IV/MPP/LABORATOR/CSHARP_TEMA/tema_sapt3/Networking/rpcprotocol/ClientRpcWorker.cs
using System;
using System.Diagnostics;
using System.Diagnostics.Eventing.Reader;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using Newtonsoft.Json;
using services;

namespace networking.rpcprotocol;

public class ClientRpcWorker:IObserver
{
    private IServices server;
    private TcpClient connection;

    private NetworkStream stream;
    private IFormatter formatter;
    private volatile bool connected;
    public ClientRpcWorker(IServices server, TcpClient connection)
    {
        this.server = server;
        this.connection = connection;
        try
        {
            stream=connection.GetStream();
            formatter = new BinaryFormatter();
            connected=true;
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }

    public virtual void run()
    {
        while(connected)
        {
            try
            {
                Object request=formatter.Deserialize(stream);
                Response response=handleRequest((Request)request);
                if (response!=null)
                {
                    sendResponse(response);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
				
            try
            {
                Thread.Sleep(1000);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }
        try
        {
            stream.Close();
            connection.Close();
        }
        catch (Exception e)
        {
            Console.WriteLine("Error "+e);
        }
    }

    
    private Response handleRequest(Request request)
    {
        Response response = null;
        switch (request.Type)
        {
            case RequestType.LOGIN:
                response=solveLogin(request);
                break;
            case RequestType.LOGOUT:
                response=solveLogout(request);
                break;
            case RequestType.GET_NUMBER_OF_PARTICIPANTS:
                response = solveVectorCnt(request);
                break;
            case RequestType.GET_PROBA_CATEGORIE_PARTICIPANTS:
                response = solveGetProbaCategorie(request);
                break;
            case RequestType.ADD_PARTICIPANT:
                response = solveInscriere(request);
                break;
        }
        return response;
    }

    
    
    private Response solveLogin(Request request)
    {
        Utilizator u = (Utilizator)request.Data;
        String username = u.username;
        String parola = u.parola;
        
        try
        {
            bool loggedIn;
            lock (server)
            {
                loggedIn = server.login(username, parola, this);
            }
            Debug.WriteLine("!!!a rezolvat loginul!!!");
            return new Response.Builder().type(ResponseType.OK).data(loggedIn).build();
            /*bool loggedIn = server.login(username, parola, this);
            return new Response.Builder().type(ResponseType.OK).data(loggedIn).build();*/
        }
        catch (Exception e)
        {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(e.Message).build();
        }
    }

    
    
    private Response solveLogout(Request request)
    {
        Utilizator u = (Utilizator) request.Data;
        String username = u.username;
        try {
            lock (server)
            {
                server.logout(username);
            }
            connected = false;
            return new Response.Builder().type(ResponseType.OK).data(true).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.Message).build();
        }
    }
    
    private Response solveVectorCnt(Request request)
    {
        try
        {
            List<int> result;
            lock (server)
            {
                result = server.creareVectorCnt();
            }
            return new Response.Builder().type(ResponseType.OK).data(result).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.Message).build();
        }
    }
    
    private Response solveGetProbaCategorie(Request request)
    {
//        System.out.println("Get all participants request");

        try {
            string proba_categorie=request.Data.ToString();
            string[] parts=proba_categorie.Split(" ");
            string proba=parts[0],categorie=parts[1];
            List<string> participanti;
            lock (server)
            {
                participanti = server.JoinInscrieriCopii(proba,categorie);
            }
            return new Response.Builder().type(ResponseType.OK).data(participanti).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.Message).build();
        }
    }
    
    private Response solveInscriere(Request request) {
        String[] inscriereData = (String[]) request.Data;

        String nume = inscriereData[0];
        var varsta = int.Parse(inscriereData[1]);
        
        List<string> probe = new List<string>();
        try {
            probe = JsonConvert.DeserializeObject<List<string>>(inscriereData[2]);
            /*foreach (var proba in probe)
            {
                Console.WriteLine(proba);
            }*/
        } catch (JsonSerializationException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.Message).build();
        }

        String usernameCurent = inscriereData[3];
        
        try {
            lock (server)
            {
                server.adaugaCopilInscriere(nume, varsta, probe,usernameCurent);
            }
            return new Response.Builder().type(ResponseType.OK).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.Message).build();
        } 
    }
    
    
    
    private void sendResponse(Response response)
    {
        Console.WriteLine("sending response "+response);
        lock (stream)
        {
            formatter.Serialize(stream, response);
            stream.Flush();
        }
    }

    
    public void participantInscris()
    {
        Response resp = new Response.Builder().type(ResponseType.UPDATE).build();
        /*System.out.println("Participant added");*/
        try
        {
            Debug.WriteLine("am trimis notificare de update");
            sendResponse(resp);
            Debug.WriteLine("a trecut de sendResponse");
        }
        catch (IOException e)
        {
            Console.WriteLine(e.Message);
        }
        catch (Exception e)
        {
            Console.WriteLine(e.Message);
        }
    }
}

File: AN_2/SEMESTRUL_IV/MPP/LABORATOR/JAVA_TEMA/tema_bun/Networking/src/main/java/com/example/tema_bun/objectProtocol/ClientRpcWorker.java
package com.example.tema_bun.objectProtocol;

import com.example.tema_bun.IObserver;
import com.example.tema_bun.IServices;
import com.example.tema_bun.Inscriere;
import com.example.tema_bun.Utilizator;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;

public class ClientRpcWorker implements Runnable, IObserver {

    private IServices server;

    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ClientRpcWorker(IServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();

            input = new ObjectInputStream(connection.getInputStream());
            connected = true;

        }catch (IOException e){
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while (connected){
            try{
                Object request = input.readObject();
                Object response = handleRequest((Request) request);
                if (response!=null){
                    sendResponse((Response) response);
                }
            }catch (IOException e){
                e.printStackTrace();
            }catch (ClassNotFoundException e){
                e.printStackTrace();
            }

            try{
                Thread.sleep(1000);
            }catch (InterruptedException e){
                e.printStackTrace();
            }
        }
        try{
            input.close();
            output.close();
            connection.close();
        }catch (IOException e){
            System.out.println("Error : " + e);
        }

    }


    private Response handleRequest(Request request){
        Response response = null;
        switch (request.type()) {
            case LOGIN: {
                response = solveLogin(request);
                break;
            }
            case LOGOUT: {
                response = solveLogout(request);
                break;
            }
            case CREARE_CONT:
            {
                response=solveAccount(request);
                break;
            }
            case GET_NUMBER_OF_PARTICIPANTS: {
                System.out.println("macar aici intra?");
                response=solveCreareVectorCnt(request);
                break;
            }
            case GET_PROBA_CATEGORIE_PARTICIPANTS:
            {
                response=solveGetProbaCategorie(request);
                break;
            }
            case INSCRIERE:
            {
                response=solveInscriere(request);
                break;
            }
//            case ADD_PARTICIPANT:
//            {
//                response=solveInscriereParticipant(request);
//                break;
//            }
        }
        return response;
    }



    private Response solveLogin(Request request) {
//        System.out.println("Login request ..." + request.type());
        Utilizator utilizator= (Utilizator) request.data();
        String username= utilizator.getUsername();
        String password = utilizator.getParola();

        try {
            Boolean loggedIn = server.login(username, password, this);
            return new Response.Builder().type(ResponseType.OK).data(loggedIn).build();
        } catch (Exception e) {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response solveAccount(Request request) {
//        System.out.println("Login request ..." + request.type());
        Utilizator utilizator= (Utilizator) request.data();
        String username= utilizator.getUsername();
        String password = utilizator.getParola();

        try {
            server.createAccount(username, password);
            Boolean loggedIn = server.login(username, password, this);
            return new Response.Builder().type(ResponseType.OK).data(loggedIn).build();
        } catch (Exception e) {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response solveLogout(Request request) {
//        System.out.println("Logout request");
        Utilizator utilizator= (Utilizator) request.data();
        String email= utilizator.getUsername();
        try {
            server.logout(email);
            connected = false;
            return new Response.Builder().type(ResponseType.OK).data(true).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response solveInscriere(Request request) {
        String[] inscriereData = (String[]) request.data();

        String nume = inscriereData[0];
        Integer varsta = Integer.parseInt(inscriereData[1]);

/*        ObjectMapper objectMapper = new ObjectMapper();
        List<String> probe = new ArrayList<>();
        try {
            probe = objectMapper.readValue(inscriereData[2], new TypeReference<List<String>>() {});
        } catch (JsonProcessingException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }*/
        Gson gson = new Gson();
        List<String> probe = new ArrayList<>();
        try {
            probe = gson.fromJson(inscriereData[2], new TypeToken<List<String>>() {}.getType());
        } catch (JsonSyntaxException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }

        String usernameCurent=inscriereData[3];

        try {
            server.adaugaCopilInscriere(nume, varsta, probe,usernameCurent);
            return new Response.Builder().type(ResponseType.OK).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }


    private Response solveCreareVectorCnt(Request request)
    {
//        System.out.println("Ger number of participants request");
        try
        {
/*            System.out.println("Pas 2");*/
            List<Integer> result=server.creareVectorCnt();
//            List<Integer> result=new ArrayList<>();
//            result.add(1);
            return new Response.Builder().type(ResponseType.OK).data(result).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response solveGetProbaCategorie(Request request)
    {
//        System.out.println("Get all participants request");

        try {
            String proba_categorie=request.data().toString();
            String[] parts=proba_categorie.split(" ");
            String proba=parts[0],categorie=parts[1];
            List<String> participanti=server.JoinInscrieriCopii(proba,categorie);
            return new Response.Builder().type(ResponseType.OK).data(participanti).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }
    private void sendResponse(Response response) throws IOException{
        System.out.println("sending response : " + response);
        synchronized (output){
            output.writeObject(response);
            output.flush();
        }
    }

    @Override
    public void participantInscris() {
        Response resp = new Response.Builder().type(ResponseType.UPDATE).build();
        System.out.println("Participant added");
        try {
            sendResponse(resp);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
